### DISCUSSION TITLE
Widgets: Click events

### DISCUSSION THREAD

**Discussion by EvanCarroll** [ratatui/ratatui#1930] [Discussion]
There was previously a feature request on rust-tui for click events. 

https://github.com/fdehau/tui-rs/issues/166

It's not clear if it's yet supported -- in which case where is documented. Or slated to be supported later. But I figured it'd be good to carry this over.

**Comment by joshka** [comment]
Definitely agree - I don't think it ended up getting written up, but would like to see it done.
The primary problems to solve with this are:
- Widgets don't have any sort of identifier (which seems like a requirement of identifying what is clicked)
- Rendering widgets doesn't track any information about their location
- Widgets that render multiple items / positions don't expose a way to map the internal location of items from position back to the item (e.g. List / Table)

The first two can be solved without any changes to Ratatui, by tracking the rects that we pass in the draw closure. That said, it might add value to do this in ratatui. I'd suggest solve it externally and then base a design on the approach.

The last probably needs some changes to those widgets, but it's somewhat difficult as the position will only be known after rendering. I think there's maybe a missing higher level concept of having widgets expose a method for layout, separate to rendering. In theory it would be possible to solve the problem without ratatui changes too, by duplicating the logic of calculating item heights.

I'd recommend:
- Implement click events in an app, note any missing pieces that make it difficult. Note particularly what the effect of implementing click events has on the structure of your code base (event based / async) and the issues of how to make this work against all the backends.
- Survey existing apps to understand patterns for how they handle this (liberally use github search and crates.io reverse dependencies to find these).
- Survey and summarize the existing issues in tui-rs / app repos.
- Document possible approaches to the problem (add some pros/cons for alternative approaches).

**Comment by joshka** [comment]
An example of an app with a custom widget that supports click events is https://github.com/ratatui-org/ratatui/blob/f767ea7d3766887cb79145103b5aa92e0eabf8f6/examples/custom_widget.rs#L241-L272

**Comment by joshka** [comment]
Closing out a bunch of old issues. This is more of a how to do / request for docs I think, so moving to discussions.
